src/System.Private.CoreLib/shared/System/Runtime/CompilerServices/AsyncMethodBuilder.cs

```diff
@@ -544,12 +544,16 @@ private class AsyncStateMachineBox<TStateMachine> : // SOS DumpAsync command dep
             where TStateMachine : IAsyncStateMachine
         {
             /// <summary>Delegate used to invoke on an ExecutionContext when passed an instance of this box type.</summary>
-            private static readonly ContextCallback s_callback = s =>
+            private static readonly ContextCallback s_callback = ExecutionContextCallback;
```

Comment:
Have you tried to move the lambda to the body? This forces static ctor which looks like is not a common path here.

Reply:
> This forces static ctor which looks like is not a common path here.

This callback is common path.  Unless you've explicitly suppressed ExecutionContext flow with ExecutionContext.SuppressFlow(), this callback will be used every time the async state machine resumes after an await.

Reply:
Also, before-field-init static ctors are executed lazily in CoreCLR so they pose no problem there. IIRC, it is different in Mono - before-field-init static ctors are eager in Mono that makes use of the before-field-init as a startup optimization ineffective.